### ISSUE TITLE
LLama 2 finetuning on multi-GPU with long context length

### ISSUE THREAD

**Issue by amant555** [huggingface/trl#804]
I am trying to train llama2 13 B model over 8 A100 80 GB. I want to train the model with 16k context length.

I have few things to get clarification on and some issues I am facing.

Things to be clarified:
1. To increase the context length I just need to set the required context length in max_seq_length parameter of SFTT?

Issues:
When model is being loaded upon accelerate launch. It does DDP rather than Tensor parallelism(TP) or Zero. Simillar things happen even if I set config with Deepseed Zero while starting the training. And since model copy is loaded on each GPU rather than it's TP. I get OOM error.

Questions related to issue:

- What is the best way to perform the training in this case?

> As if I use Naive model parallelism(MP) I will have a bubble where all GPU are not utilised fully and I am not able to use MP with distributed manner where that same MP model is replicated on another set of GPUs.

So, Is there a way I can have a TP/PP model and do distributed training and what is the best way to implement?
As for MP i have to set device map with max memory it can use to load model.

**Comment by Ricardokevins** [comment]
> I am trying to train llama2 13 B model over 8 A100 80 GB. I want to train the model with 16k context length.
> 
> I have few things to get clarification on and some issues I am facing.
> 
> Things to be clarified:
> 
> 1. To increase the context length I just need to set the required context length in max_seq_length parameter of SFTT?
> 
> Issues: When model is being loaded upon accelerate launch. It does DDP rather than Tensor parallelism(TP) or Zero. Simillar things happen even if I set config with Deepseed Zero while starting the training. And since model copy is loaded on each GPU rather than it's TP. I get OOM error.
> 
> Questions related to issue:
> 
> * What is the best way to perform the training in this case?
> 
> > As if I use Naive model parallelism(MP) I will have a bubble where all GPU are not utilised fully and I am not able to use MP with distributed manner where that same MP model is replicated on another set of GPUs.
> 
> So, Is there a way I can have a TP/PP model and do distributed training and what is the best way to implement? As for MP i have to set device map with max memory it can use to load model.

Hi , do you full tune the model or just LORA?
when i try to ppo full-tune a 7B model (512 length), I got OOM on A100 80G. So i am wondering if there is some problem in my stetting or not?

**Comment by younesbelkada** [comment]
Hi @Ricardokevins @amant555 
You can now use Flash Attention 1 & 2 to overcome OOM issues, please see: https://huggingface.co/docs/trl/main/en/sft_trainer#using-flash-attention-and-flash-attention-2 for more details

**Comment by Ricardokevins** [comment]
> Hi @Ricardokevins @amant555 You can now use Flash Attention 1 & 2 to overcome OOM issues, please see: https://huggingface.co/docs/trl/main/en/sft_trainer#using-flash-attention-and-flash-attention-2 for more details

thank you for your constructive suggestions!
is this method is also valid in PPO stage ( I saw this method is introduced in SFTtrainer)

**Comment by younesbelkada** [comment]
if you pass an `AutoModelForCausalLM` to `AutoModelForCausalLMWithValueHead` it should work yes, but only for FA2 as FA1 does not support training with padding tokens

**Comment by amant555** [comment]
> Hi @Ricardokevins @amant555 You can now use Flash Attention 1 & 2 to overcome OOM issues, please see: https://huggingface.co/docs/trl/main/en/sft_trainer#using-flash-attention-and-flash-attention-2 for more details

Thanks @younesbelkada I will try this. But for the purpose of the fine-tuning how can I better parallelise the model. For faster training.

I have followed and did the normal pip install of each packages such as deepseed and others. 

Update:

When I now run the trainer in Naive model parallel it works but I have to run gardient accumulation at 1. Which I don't think is good. Any suggestions what can be done?
